### PR TITLE
Implement max series limit in distributor for `/series` API

### LINKED ISSUE #4667: `max_fetched_series_per_query` limits are not respected for `/series` API for recent data
**Describe the bug**
GetSeries and GetLabels API can retrieve informations from Ingesters or/and storegateways (when `query_store_for_labels_enabled` configuration is enabled).

We have different behaviors if the requests are going to `Ingesters` and to `storegateways`.
On the first case, the `max_fetched_series_per_query` is not enforced at all while in latter case this configuration is enforced:

StoreGateway: https://github.com/cortexproject/cortex/blob/45f043f0d121a04a4da0d60e7791562f93c6aa91/pkg/querier/blocks_store_queryable.go#L625

Ingester:
https://github.com/cortexproject/cortex/blob/45f043f0d121a04a4da0d60e7791562f93c6aa91/pkg/distributor/distributor.go#L958

This behavior can cause OOM on the queriers as there is a unbounded limit for number of series brought from ingesters.

**To Reproduce**
1. Start Cortex (SHA or version) : any
2. Configure `query_store_for_labels_enabled` to 10
3. Send more than 10 series
4. call `/series` with `match[]={__name__=~".+"}`

**Expected behavior**
`max_fetched_series_per_query` limit should be respected on both cases (data coming from StoreGateway or Ingester)

**Storage Engine**
- [X] Blocks
- [ ] Chunks

**Additional Context**

### PR DESCRIPTION
**Pull request by harry671003**
Signed-off-by: 🌲 Harry 🌊 John 🏔 <johrry@amazon.com>

**What this PR does**:
This PR enforces the `max_fetched_series_per_query` limit for `/api/v1/series` API.

**Which issue(s) this PR fixes**:
Fixes #4667

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/distributor_queryable.go [hunk 1]
```diff
@@ -97,10 +98,22 @@ func (q *distributorQuerier) Select(_ bool, sp *storage.SelectHints, matchers ..
 	// See: https://github.com/prometheus/prometheus/pull/8050
 	if sp != nil && sp.Func == "series" {
 		ms, err := q.distributor.MetricsForLabelMatchers(ctx, model.Time(q.mint), model.Time(q.maxt), matchers...)
+		queryLimiter := limiter.QueryLimiterFromContextWithFallback(ctx)
```
Comment (alanprot) [inline pkg/querier/distributor_queryable.go:101]: I think this change is easier if we do inside the `MetricsForLabelMatchers`:

https://github.com/aws-observability/cortex-dev/blob/fc862f9c387bbc4f615d79d1a7854d0d3d21aa89/pkg/distributor/distributor.go#L973-L976

Another possible memory optimization on this code would be to not wait all `resps` and dedup those resps as soon as it arrive inside the `ForReplicationSet`

Comment (harry671003) [reply pkg/querier/distributor_queryable.go]: Thanks for the feedback.

**Comment by alanprot** [comment]
LGTM

**Review by alanprot — APPROVED** [review]

File: pkg/distributor/distributor_test.go [hunk 2]
```diff
@@ -1924,6 +1936,19 @@ func TestDistributor_MetricsForLabelMatchers(t *testing.T) {
 				{Metric: util.LabelsToMetric(fixtures[1].lbls)},
 			},
 			expectedIngesters: numIngesters,
+			queryLimiter:      limiter.NewQueryLimiter(0, 0, 0),
+			expectedErr:       nil,
+		},
+		"should return err if series limit is exhausted": {
```
[the remark this answers is not in the document]
Comment (harry671003) [reply pkg/distributor/distributor_test.go:1942]: Thanks for reviewing. The [replication factor is 3](https://github.com/cortexproject/cortex/blob/master/pkg/distributor/distributor_test.go#L2109) by default. I added a test to cover the case where one series is queried with limit is set to 1.

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -953,20 +955,29 @@ func (d *Distributor) MetricsForLabelMatchers(ctx context.Context, from, through
...
+		resp, err := client.MetricsForLabelMatchers(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		ms := ingester_client.FromMetricsForLabelMatchersResponse(resp)
 		for _, m := range ms {
+			if err := queryLimiter.AddSeries(cortexpb.FromMetricsToLabelAdapters(m)); err != nil {
+				return nil, err
+			}
+			mutex.Lock()
```
[the remark this answers is not in the document]
Comment (harry671003) [reply pkg/distributor/distributor.go:971]: I added a benchmark test `BenchmarkDistributor_MetricsForLabelMatchers`. 

#### Without the locking:
```
❯ go test -run=^$ -bench ^BenchmarkDistributor_MetricsForLabelMatchers$ github.com/cortexproject/cortex/pkg/distributor -count 10
goos: darwin
goarch: amd64
pkg: github.com/cortexproject/cortex/pkg/distributor
cpu: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     775	   1553236 ns/op	  758952 B/op	    8580 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     724	   1551052 ns/op	  759031 B/op	    8581 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     898	   1281153 ns/op	  759056 B/op	    8581 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	    1064	   1145841 ns/op	  758964 B/op	    8580 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	    1102	   1074635 ns/op	  759169 B/op	    8581 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	    1123	   1068547 ns/op	  759050 B/op	    8580 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	    1108	   1080608 ns/op	  759064 B/op	    8581 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	    1129	   1052474 ns/op	  758646 B/op	    8578 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	    1129	   1123325 ns/op	  758817 B/op	    8579 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	    1084	   1360482 ns/op	  758859 B/op	    8580 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/distributor	14.253s
```


#### With the locking
```
❯ go test -run=^$ -bench ^BenchmarkDistributor_MetricsForLabelMatchers$ github.com/cortexproject/cortex/pkg/distributor -count 10
goos: darwin
goarch: amd64
pkg: github.com/cortexproject/cortex/pkg/distributor
cpu: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     847	   1519934 ns/op	  870517 B/op	    9174 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     804	   1789515 ns/op	  870984 B/op	    9178 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     542	   2008873 ns/op	  871077 B/op	    9179 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     715	   1493901 ns/op	  871080 B/op	    9179 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     793	   1619484 ns/op	  871042 B/op	    9179 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     712	   1502533 ns/op	  870682 B/op	    9176 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     872	   1594814 ns/op	  870655 B/op	    9176 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     652	   1608638 ns/op	  870481 B/op	    9175 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     806	   1452676 ns/op	  870680 B/op	    9176 allocs/op
BenchmarkDistributor_MetricsForLabelMatchers/get_series_within_limits-8         	     681	   1634886 ns/op	  870830 B/op	    9177 allocs/op
PASS
ok  	github.com/cortexproject/cortex/pkg/distributor	15.305s
```

I didn't see any significant differences.

**Pull request merged**